src/Mod/Sketcher/Gui/ViewProviderSketch.cpp

```diff
@@ -3899,4 +3816,153 @@ bool ViewProviderSketch::isInEditMode() const
 {
     return editCoinManager != nullptr;
 }
+void ViewProviderSketch::generateContextMenu() {
+    int selectedEdges = 0;
+    int selectedLines = 0;
+    int selectedConics = 0;
+    int selectedPoints = 0;
+    int selectedConstraints = 0;
+
+    Gui::MenuItem geom;
+    geom.setCommand("Sketcher context");
+
+    std::vector<Gui::SelectionObject> selection = Gui::Selection().getSelectionEx(0, Sketcher::SketchObject::getClassTypeId());
+
+    if (selection.size() > 0) {
+        const std::vector<std::string> SubNames = selection[0].getSubNames();
+
+        for (auto name : SubNames) {
+            if (name.substr(0, 4) == "Edge") {
+                ++selectedEdges;
+
+                int geoId = std::atoi(name.substr(4, 4000).c_str()) - 1;
+                if (geoId >= 0) {
+                    const Part::Geometry* geo = getSketchObject()->getGeometry(geoId);
+                    if (isLineSegment(*geo)) {
+                        ++selectedLines;
+                    }
+                    else {
+                        ++selectedConics;
+                    }
+                }
+            }
+            else if (name.substr(0, 4) == "Vert") {
+                ++selectedPoints;
+            }
+            else if (name.substr(0, 4) == "Cons") {
+                ++selectedConstraints;
+            }
+        }
+
+        if (selectedEdges >= 1 && selectedPoints == 0) {
+            // Here you want to make special cases for line vs arcs vs circle and so on.
+            geom << "Sketcher_Dimension";
+            if (selectedConics == 0) {
+                geom << "Sketcher_ConstrainHorVer"
+                << "Sketcher_ConstrainVertical"
+                << "Sketcher_ConstrainHorizontal";
+            }
+            else if (selectedConics > 1 && selectedLines == 0){
+                geom << "Sketcher_ConstrainCoincidentUnified"
+                    << "Sketcher_ConstrainTangent"
+                    << "Sketcher_ConstrainEqual";
+            }
+            else if (selectedConics == 1 && selectedLines == 1){
+                geom << "Sketcher_ConstrainTangent";
+            }
+            if (selectedLines > 1 && selectedConics == 0) {
+                geom << "Sketcher_ConstrainParallel"
+                    << "Sketcher_ConstrainPerpendicular"
+                    << "Sketcher_ConstrainTangent"
+                    << "Sketcher_ConstrainEqual";
+            }
+        }
+        else if (selectedEdges == 1 && selectedPoints >= 1) {
+            // Here you want to make special cases for line vs arcs vs circle and so on.
+            geom << "Sketcher_Dimension";
+            if (selectedConics == 0) {
+                geom << "Sketcher_ConstrainCoincidentUnified"
+                    << "Sketcher_ConstrainHorVer"
+                    << "Sketcher_ConstrainVertical"
+                    << "Sketcher_ConstrainHorizontal";
+                if (selectedPoints == 2) {
+                geom << "Sketcher_ConstrainSymmetric";
```

Comment:
Try to reformat new code using clang-format

Reply:
Or if you are using visual studio, you can also cut the whole paragraph and paste it back and visual studio should fix the indentation automatically on paste.